src/brod_consumer.erl

```diff
@@ -48,238 +42,209 @@
+    proplists:get_value(sleep_timeout, Config, ?DEFAULT_SLEEP_TIMEOUT),
+  SleepTimeout =
+    proplists:get_value(sleep_timeout, CbOptions, SleepTimeout0),
+  Offset0 = proplists:get_value(offset, Config, ?DEFAULT_OFFSET),
+  Offset1 = proplists:get_value(offset, CbOptions, Offset0),
+
+  %% 3. Retrive valid starting offset
+  case get_valid_offset(SocketPid, Offset1, Topic, Partition) of
     {error, Error} ->
-      {reply, {error, Error}, State0};
+      {stop, {error, Error}, State0};
     {ok, Offset} ->
-      State = State0#state{ callback      = Callback
+      State = State0#state{ socket_pid    = SocketPid
                           , offset        = Offset
                           , max_wait_time = MaxWaitTime
                           , min_bytes     = MinBytes
-                          , max_bytes     = MaxBytes},
-      ok = send_fetch_request(State),
-      {reply, ok, State}
+                          , max_bytes     = MaxBytes
+                          , sleep_timeout = SleepTimeout
+                          },
+      %% 4. Start consuming
+      {ok, CorrId} = send_fetch_request(State),
+      {noreply, State#state{corr_id = CorrId}}
   end;
-handle_call(stop, _From, State) ->
-  {stop, normal, ok, State};
-handle_call(get_socket, _From, State) ->
-  {reply, {ok, State#state.socket}, State};
-handle_call(Request, _From, State) ->
-  {stop, {unsupported_call, Request}, State}.
-
-handle_cast(Msg, State) ->
-  {stop, {unsupported_cast, Msg}, State}.
-
-handle_info({msg, _Pid, _CorrId, #fetch_response{} = R}, State0) ->
-  case handle_fetch_response(R, State0) of
-    {error, Error} ->
-      {stop, {broker_error, Error}, State0};
-    {ok, State} ->
-      MessageSet = brod_utils:fetch_response_to_message_set(R),
-      exec_callback(State#state.callback, MessageSet),
-      ok = send_fetch_request(State),
-      {noreply, State};
-    {empty, State} ->
-      erlang:send_after(State#state.sleep_timeout, self(), ?SEND_FETCH_REQUEST),
-      {noreply, State}
+handle_info({msg, _Pid, CorrId, R}, #state{corr_id = CorrId} = State0) ->
+  #fetch_response{topics = [TopicFetchData]} = R,
+  #topic_fetch_data{ topic = Topic
+                   , partitions = [PM]} = TopicFetchData,
+  #partition_messages{ partition = Partition
+                     , error_code = ErrorCode
+                     , high_wm_offset = HighWmOffset
+                     , last_offset = LastOffset
+                     , messages = Messages} = PM,
+  case brod_kafka:is_error(ErrorCode) of
+    true ->
+      %% TODO: do something smarter
+      %% 1) handle what we can handle
+      %% 2) call CbMod:handle_error for errors we can not handle
+      %%    it can for example bump max_bytes option or skip a message
+      %% 3) stop if CbMod:handle_error is not defined
+      {stop, {broker_error, ErrorCode}, State0};
+    false ->
+      case Messages of
+        [] ->
+          erlang:send_after(State0#state.sleep_timeout, self(), ?SEND_FETCH_REQUEST),
```

Comment:
self() ! ?SEND_FETCH_REQUEST directly if sleep_timeout is zero.
so it doesn't have to schedule a timer